sdk/python/kfp/registry/client.py

```diff
@@ -0,0 +1,326 @@
+            config = {
+                'host':
+                    self._host,
+                'upload_url':
+                    self._host,
+                'download_version_url':
+                    f'{self._host}/{{package_name}}/{{version}}',
+                'download_tag_url':
+                    f'{self._host}/{{package_name}}/{{tag}}',
+                'get_package_url':
+                    f'{package_name_endpoint}',
+                'list_packages_url':
+                    package_endpoint,
+                'delete_package_url':
+                    f'{package_name_endpoint}',
+                'get_tag_url':
+                    f'{tags_endpoint}/{{tag}}',
+                'list_tags_url':
+                    f'{tags_endpoint}',
+                'delete_tag_url':
+                    f'{tags_endpoint}/{{tag}}',
+                'create_tag_url':
+                    f'{tags_endpoint}?tagId={{tag}}',
+                'update_tag_url':
+                    f'{tags_endpoint}/{{tag}}?updateMask=version',
+                'get_version_url':
+                    f'{versions_endpoint}/{{version}}',
+                'list_versions_url':
+                    f'{versions_endpoint}',
+                'delete_version_url':
+                    f'{versions_endpoint}/{{version}}',
+                'package_format':
+                    f'{repo_resource_format}/packages/{{package_name}}',
+                'tag_format':
+                    f'{repo_resource_format}/packages/{{package_name}}/tags/{{tag}}',
+                'version_format':
+                    f'{repo_resource_format}/packages/{{package_name}}/versions/{{version}}',
+            }
+        else:
+            logging.info(f'load_config not implemented for host: {self._host}')
+        return config
+
+    def _get_auth(self):
+        auth = self._auth
+        if isinstance(auth, Credentials):
+            auth = ApiAuth(auth.token)
+        return auth
+
+    def _refresh_creds(self):
+        if self._is_ar_host() and isinstance(
+                self._auth, Credentials) and not self._auth.valid:
+            self._auth.refresh(google.auth.transport.requests.Request())
+
+    def upload_pipeline(self, file_name: str, tags: Optional[Union[str,
+                                                                   List[str]]],
+                        extra_headers: Optional[dict]) -> Tuple[str, str]:
+        url = self._config['upload_url']
+        self._refresh_creds()
+        auth = self._get_auth()
+        request_body = {}
+        if tags:
+            if isinstance(tags, str):
+                request_body = {'tags': tags}
+            elif isinstance(tags, List):
+                request_body = {'tags': ','.join(tags)}
+
+        files = {'content': open(file_name, 'rb')}
+        response = requests.post(
+            url=url,
+            data=request_body,
+            headers=extra_headers,
+            files=files,
```

Comment:
consider: passing `**kwargs` from `self._requests` could allow you to neatly wrap this call into the same `self._requests`

Reply:
I think the code is clearer by passing the variables explicitly. I'll leave this as-is for now.